core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1739,18 +1756,24 @@ private[spark] class DAGScheduler(
           if (mapStage.rdd.isBarrier()) {
             // Mark all the map as broken in the map stage, to ensure retry all the tasks on
             // resubmitted stage attempt.
-            mapOutputTracker.unregisterAllMapOutput(shuffleId)
+            mapOutputTracker.unregisterAllMapAndMergeOutput(shuffleId)
```

Comment:
Similar to the handling of indeterminate stage retry, should we also reset other metadata here?

Reply:
Same as above, added a comment here to handle it as part of handling barrier execution mode handling.